### PR TITLE
Cpp codegen: handling of imported exceptions. Fixes #3571

### LINKED ISSUE #3571: Catching C++ exceptions in Nim
I'm trying to wrap a C++ library that can through a ton of exceptions, and it looks like there's not a way to catch C++ exceptions from within Nim code.

Here is the Nim code with my failed attempt to wrap C++ exceptions:

``` nim
{.compile: "PingC.cpp".}

import os

const pingc = "/source/PingC.h"

type
  CorbaOrbVar {.final, header: pingc, importcpp: "CORBA::ORB_var".} = object
  CorbaObjectVar {.final, header:pingc, importcpp: "CORBA::Object_var".} =
    object
  ExamplePingableVar {.final, header:pingc,
                       importcpp: "Example::Pingable_var".} = object
  CorbaTransient {.final, header:pingc, importcpp: "CORBA::TRANSIENT".} = object of Exception

proc corbaOrbInit(argc: cint, argv: cstringArray): CorbaOrbVar
  {.header: pingc, importcpp: "CORBA::ORB_init(@)".}

proc stringToObject(orb: CorbaOrbVar, s: cstring): CorbaObjectVar
  {.header: pingc, importcpp: "#->string_to_object(@)".}

proc destroy(orb: CorbaOrbVar)
  {.header: pingc, importcpp: "#->destroy(@)".}

proc examplePingableNarrow(obj: CorbaObjectVar): ExamplePingableVar
  {.header: pingc, importcpp: "Example::Pingable::_narrow(@)".}

proc ping(pingable: ExamplePingableVar)
  {.header: pingc, importcpp: "#->ping(@)", raises: [CorbaTransient].}

var params = os.commandLineParams()
var argc: cint = params.len().cint
var argv: cstringArray = allocCStringArray(params)

try:
  var orb = corbaOrbInit(argc, argv)
  var po = orb.stringToObject("corbaname::nameservice/NameService#test/ExamplePing")
  var pingable = examplePingableNarrow(po)
  pingable.ping()
  orb.destroy()
except CorbaTransient:
  echo("Got an exception!")
```

I tried wrapping the exception I wanted to catch and having it inherit from `system.Exception`. I've also declared the wrapped function that can throw that exception with the `{.raises.}` pragma. However, I still can't catch the exception.

Here's what the compiled C++ looks like:

``` cpp
        try {
                nimln(38, "tao_ping_client.nim");
                CORBA::ORB_var orb = CORBA::ORB_init(argc_140063, argv_140068);
                nimln(39, "tao_ping_client.nim");
                CORBA::Object_var po = orb->string_to_object("corbaname::nameservice/NameService#test/ExamplePing");
                nimln(40, "tao_ping_client.nim");
                Example::Pingable_var pingable = Example::Pingable::_narrow(po);
                nimln(41, "tao_ping_client.nim");
                pingable->ping();
                nimln(42, "tao_ping_client.nim");
                orb->destroy();
        } catch (NimException& TMP526) {
        setFrame((TFrame*)&FR);
        if (isObj(TMP526.exp->m_type, (&NTI140015)))    {
                nimln(44, "tao_ping_client.nim");
                printf("%s\012", ((NimStringDesc*) &TMP527)? (((NimStringDesc*) &TMP527))->data:"nil");
                popCurrentException();
        }
        else    {
                throw;
        }
        }
```

There isn't a catch block for the type I'm trying to catch, which doesn't derive from `NimException`. I think the compiler should check to see if any of the `except` blocks are catching a type with the `{.importcpp.}` pragma, and add an appropriate catch block for it in the generated C++.

### PR DESCRIPTION
**Pull request by cooldome**
PR implements throwing and catching of C++ exceptions. Any type with ``importcpp`` pragma can be raised or caught. ~~The PR is lacking only a changelog entry and a documentation update.~~

In order to understand the design hacks made, please consider the following use case:
```
try {}
catch(...)
{
// nothing is known about the caught type, but if it was a Nim type 
// then popCurrentException() call is required 
}
```
There is only one way to get ``popCurrentException()`` called in such case  without pushing burden to the user. It is to put ``popCurrentException() `` call into destructor of Nim's Exception type and start throwing exceptions by value and catching by reference.
The use cpp of destructors has also one big benefit: early return statements like return/break cleanup automatically which is nice.

One complication found, the cpp destructors can be called in order other than ``pushCurrentException`` was called hence I have added support to pop exceptions from the exception stack by ``raise_id`` identifier.

### CODE REVIEW

File: compiler/cgen.nim [hunk 1]
```diff
@@ -14,7 +14,7 @@ import
   nversion, nimsets, msgs, std / sha1, bitsets, idents, types,
   ccgutils, os, ropes, math, passes, rodread, wordrecg, treetab, cgmeth,
   condsyms, rodutils, renderer, idgen, cgendata, ccgmerge, semfold, aliases,
-  lowerings, semparallel, tables, sets, ndi
+  lookups, lowerings, semparallel, tables, sets, ndi
```
Comment (Araq) [inline compiler/cgen.nim]: Why is this required? ``lookups`` is for the frontend only.

File: compiler/sempass2.nim [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 import
   intsets, ast, astalgo, msgs, renderer, magicsys, types, idents, trees,
-  wordrecg, strutils, options, guards, writetracking
+  wordrecg, strutils, options, guards, writetracking, lookups
```
Comment (Araq) [inline compiler/sempass2.nim]: I think ``isInfixAs`` should not be in ``lookups``.

File: compiler/semstmts.nim [hunk 3]
```diff
@@ -286,18 +292,24 @@ proc semTry(c: PContext, n: PNode): PNode =
...
         # support ``except KeyError, ValueError, ... : body``
+        var is_native, is_imported: bool
         for j in 0..a.len-2:
-          a[j] = semExceptBranchType(a[j])
+          let tmp = semExceptBranchType(a[j])
+          if tmp: is_imported = true
+          else: is_native = true
+
+        if is_native and is_imported:
+          local_error(a[0].info, "Mix of imported and native exception types is not allowed in one except branch")
```
Comment (Araq) [inline compiler/semstmts.nim]: That's spelt ``localError``

File: compiler/semstmts.nim [hunk 4]
```diff
@@ -731,16 +743,21 @@ proc semFor(c: PContext, n: PNode): PNode =
...
-    var typ = n.sons[0].typ
-    if typ.kind != tyRef or typ.lastSon.kind != tyObject:
-      localError(n.info, errExprCannotBeRaised)
-
-    # check if the given object inherits from Exception
-    if not typ.lastSon.isException():
-        localError(n.info, "raised object of type $1 does not inherit from Exception",
-                           [typeToString(typ)])
+  if n[0].kind == nkEmpty:
+    return
```
Comment (Araq) [inline compiler/semstmts.nim]: not a fan of early returns.

File: lib/system/excpt.nim [hunk 5]
```diff
@@ -320,6 +337,11 @@ proc raiseExceptionAux(e: ref Exception) =
       quitOrDebug()
     else:
       pushCurrentException(e)
+      discard popCurrentExceptionEx # small hack to trigger dependency for codegen
```
Comment (Araq) [inline lib/system/excpt.nim]: There must be a clean(er) way to do this.

File: compiler/sempass2.nim (hunk 2 above)
Comment (cooldome) [reply compiler/sempass2.nim]: moved to ast.nim

**Comment by cooldome** [comment]
transf rewrites isInfixAs for Nim exceptions, but keeps imported exceptions untouched. I was thinking if it is possible to use ``catch(KeyError& e)`` for Nim exceptions as well, but cpp exceptions are thrown by value for destructors to work while Nim wants it to be by reference therefore `transf` is pretty useful even for cpp backend

File: lib/system/excpt.nim (hunk 5 above)
Comment (cooldome) [reply lib/system/excpt.nim]: The problem is that I can't emit ``popCurrentExceptionEx`` dependency during ``getTypeDesc`` call for ``Exception`` type as ``Exception`` type is used inside ``popCurrentExceptionEx``. Endless loop.

The only idea I have is to introduce something like ``TCGen.typeStack``, but for procs/syms so I can push ``popCurrentExceptionEx`` into it and the proc gets emitted in ``writeModule`` as finalization step

More ideas welcome

File: compiler/semstmts.nim (hunk 3 above)
Comment (cooldome) [reply compiler/semstmts.nim]: fixed

File: compiler/semstmts.nim (hunk 4 above)
Comment (cooldome) [reply compiler/semstmts.nim]: done, but now it is many nested if statements which is not good either

**Comment by cooldome** [comment]
Ready for the next review

File: compiler/ccgstmts.nim [hunk 6]
```diff
@@ -834,8 +834,12 @@ proc genTryCpp(p: BProc, t: PNode, d: var TLoc) =
       endBlock(p)
     else:
       for j in 0..t[i].len-2:
-        assert(t[i][j].kind == nkType)
-        startBlock(p, "catch ($1*) {$n", getTypeDesc(p.module, t[i][j].typ))
+        if t[i][j].isInfixAs():
+          let v = rope($t[i][j][2])
```
Comment (Araq) [inline compiler/ccgstmts.nim]: wait ... what? Pretty sure that's not how to generate C++ code.

File: compiler/ccgstmts.nim [hunk 7]
```diff
@@ -834,8 +834,12 @@ proc genTryCpp(p: BProc, t: PNode, d: var TLoc) =
...
+          fillLoc(t[i][j][2].sym.loc, locTemp, t[i][j][2], v, OnUnknown)
```
Comment (Araq) [inline compiler/ccgstmts.nim]: Some name for ``t[i][j][2]`` would be nice. (Yes, I think I know what it is.)

File: lib/system/excpt.nim [hunk 8]
```diff
@@ -320,6 +337,10 @@ proc raiseExceptionAux(e: ref Exception) =
       quitOrDebug()
     else:
       pushCurrentException(e)
+      raise_counter.inc
+      if raise_counter == 0:
+        raise_counter.inc # skip zero at overflow
```
Comment (Araq) [inline lib/system/excpt.nim]: seriously? You raise the same exception more than 4 billion times? Doesn't hell freeze over before that happens in practice?

File: compiler/ccgstmts.nim (hunk 7 above)
Comment (cooldome) [reply compiler/ccgstmts.nim]: done

File: compiler/ccgstmts.nim (hunk 6 above)
Comment (cooldome) [reply compiler/ccgstmts.nim]: Agree, fixed in the last version

**Comment by cooldome** [comment]
Ready for one more review, added changelog entry and a section to the language manual

File: changelog.md [hunk 9]
```diff
@@ -37,6 +37,10 @@
   the use of `static[T]` types.
   (#6415)
 
+- Native C++ exceptions can now be imported with `importcpp` pragma. 
+  Imported exceptions can be raised and caught just like Nim exceptionы.
```
Comment (Yardanico) [inline changelog.md]: "exceptionЫ" - cyrillic "Ы" instead of latin "s" :) offtop: I know it can be pain sometimes to switch keyboard layouts

Comment (Yardanico) [reply changelog.md]: Oh, already fixed, nvm

Comment (cooldome) [reply changelog.md]: yep, switch is a real pain if you have 3+ keyboard layouts.
Fixed

**Comment by Araq** [comment]
Ok, now looks excellent, Thank you!

**Pull request merged**